holoviews/plotting/plot.py

```diff
@@ -825,44 +858,49 @@ def _compute_group_range(cls, group, elements, ranges, framewise, robust, top_le
             if matching:
                 group_dim_ranges[gdim] = values
 
+        # Merge ranges across elements
         dim_ranges = []
         for gdim, values in group_dim_ranges.items():
-            hard_range = util.max_range(values['hard'], combined=False)
-            soft_range = util.max_range(values['soft'])
-            robust_range = util.max_range(values.get('robust', []))
-            data_range = util.max_range(values['data'])
-            combined = util.dimension_range(data_range[0], data_range[1],
-                                            hard_range, soft_range)
-            dranges = {'data': data_range, 'hard': hard_range,
-                       'soft': soft_range, 'combined': combined,
-                       'robust': robust_range}
-            if 'factors' in values:
-                all_factors = values['factors']
-                factor_dtypes = {fs.dtype for fs in all_factors} if all_factors else []
-                dtype = list(factor_dtypes)[0] if len(factor_dtypes) == 1 else None
-                expanded = [v for fctrs in all_factors for v in fctrs]
-                if dtype is not None:
-                    try:
-                        # Try to keep the same dtype
-                        expanded = np.array(expanded, dtype=dtype)
-                    except Exception:
-                        pass
-                dranges['factors'] = util.unique_array(expanded)
+            dranges = cls._merge_group_ranges(values)
             dim_ranges.append((gdim, dranges))
-        if prev_ranges and not (framewise and top_level):
+
+        # Merge local ranges into global range dictionary
```

Comment:
This is the hardest bit to follow. As I find this hard to reason about the best thing to do is make sure there are lots of  decent unit tests!